packages/core-data/src/batch/create-batch.js

```diff
@@ -0,0 +1,189 @@
+
+	return {
+		/**
+		 * Adds an input to the batch and returns a promise that is resolved or
+		 * rejected when the input is processed by `batch.run()`.
+		 *
+		 * You may also pass a thunk which allows inputs to be added
+		 * asychronously.
+		 *
+		 * ```
+		 * // Both are allowed:
+		 * batch.add( { path: '/v1/books', ... } );
+		 * batch.add( ( add ) => add( { path: '/v1/books', ... } ) );
+		 * ```
+		 *
+		 * If a thunk is passed, `batch.run()` will pause until either:
+		 *
+		 * - The thunk calls its `add` argument, or;
+		 * - The thunk returns a promise and that promise resolves, or;
+		 * - The thunk returns a non-promise.
+		 *
+		 * @param {any|Function} inputOrThunk Input to add or thunk to execute.
+		 
+		 * @return {Promise|any} If given an input, returns a promise that
+		 *                       is resolved or rejected when the batch is
+		 *                       processed. If given a thunk, returns the return
+		 *                       value of that thunk.
+		 */
+		add( inputOrThunk ) {
+			const id = ++lastId;
+			pending.add( id );
+
+			const add = ( input ) =>
+				new Promise( ( resolve, reject ) => {
+					queue.push( {
+						input,
+						resolve,
+						reject,
+					} );
+					pending.delete( id );
+				} );
+
+			if ( isFunction( inputOrThunk ) ) {
+				return Promise.resolve( inputOrThunk( add ) ).finally( () => {
+					pending.delete( id );
+				} );
+			}
+
+			return add( inputOrThunk );
+		},
+
+		/**
+		 * Runs the batch. This calls `batchProcessor` and resolves or rejects
+		 * all promises returned by `add()`.
+		 *
+		 * @return {Promise} A promise that resolves to a boolean that is true
+		 *                   if the processor returned no errors.
+		 */
+		async run() {
+			if ( pending.size ) {
+				await new Promise( ( resolve ) => {
+					const unsubscribe = pending.subscribe( () => {
+						if ( ! pending.size ) {
+							unsubscribe();
+							resolve();
+						}
+					} );
+				} );
+			}
+
+			let results;
+
+			try {
+				results = await processor(
+					queue.map( ( { input } ) => input )
+				);
+
+				if ( results.length !== queue.length ) {
+					throw new Error(
+						'run: Array returned by processor must be same size as input array.'
+					);
+				}
+			} catch ( error ) {
+				for ( const { reject } of queue ) {
+					reject( error );
+				}
+
+				throw error;
+			}
+
+			let isSuccess = true;
+
+			for ( const [ result, { resolve, reject } ] of zip(
+				results,
+				queue
+			) ) {
+				if ( result?.error ) {
+					reject( result.error );
+					isSuccess = false;
+				} else {
+					resolve( result?.output ?? result );
+				}
+			}
+
+			return isSuccess;
+		},
+	};
+}
+
+class ObservableSet {
+	constructor( ...args ) {
+		this.set = new Set( ...args );
+		this.subscribers = [];
+	}
+
+	get size() {
+		return this.set.size;
+	}
+
+	add( ...args ) {
+		this.set.add( ...args );
+		for ( const subscriber of this.subscribers ) {
+			subscriber();
+		}
```

Comment:
A-ha, another place using observables – we're getting closer to using Rx.js or kefir at some point :-)

Reply:
I had this inlined but split it out into `ObservableSet` because I think it's easier to follow when `createBatch` has less state in it to manage. Plus, this is future proof in that we can delete `ObservableSet` if we ever add an observable library to Gutenberg 🙂